MagazineLayout/LayoutCore/LayoutState.swift

```diff
@@ -0,0 +1,174 @@
+// Created by Bryan Keller on 10/31/25.
+// Copyright © 2025 Airbnb Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+
+// MARK: - LayoutState
+
+/// Represents the state of the layout, including metrics and the current `ModelState`.
+struct LayoutState {
+
+  // MARK: Internal
+
+  let modelState: ModelState
+
+  var bounds: CGRect
+  var contentInset: UIEdgeInsets
+  var scale: CGFloat
+  var verticalLayoutDirection: MagazineLayoutVerticalLayoutDirection
+
+  var minContentOffset: CGPoint {
+    CGPoint(x: -contentInset.left, y: -contentInset.top)
+  }
+
+  var maxContentOffset: CGPoint {
+    let x = contentSize.width - bounds.width + contentInset.right
+    let y = contentSize.height - bounds.height + contentInset.bottom
+    return CGPoint(x: max(x, minContentOffset.x), y: max(y, minContentOffset.y))
+  }
+
+  var contentSize: CGSize {
+    // This is a workaround for `layoutAttributesForElementsInRect:` not getting invoked enough
+    // times if `collectionViewContentSize.width` is not smaller than the width of the collection
+    // view, minus horizontal insets. This results in visual defects when performing batch
+    // updates. To work around this, we subtract 0.0001 from our content size width calculation;
+    // this small decrease in `collectionViewContentSize.width` is enough to work around the
+    // incorrect, internal collection view `CGRect` checks, without introducing any visual
+    // differences for elements in the collection view.
+    // See https://openradar.appspot.com/radar?id=5025850143539200 for more details.
+    let width = bounds.width - contentInset.left - contentInset.right - 0.0001
+
+    let numberOfSections = modelState.numberOfSections
+    let height: CGFloat =
+      if numberOfSections <= 0 {
+        0
+      } else {
+        modelState.sectionMaxY(forSectionAtIndex: numberOfSections - 1)
+      }
+
+    return CGSize(width: width, height: height)
```

Comment:
all of this used to live in `MagazineLayout.swift` - I moved it here without changing any math or comments